### PR TITLE
ADD: Feature/automatic doc deployment

### LINKED ISSUE #145: Extend automatic documentation
Add all the route information to the documentation.
Is there an automated way to include all the snippets?

### PR DESCRIPTION
**Pull request by Segelzwerg**
Documentation is automatically deployed to the github pages upon pushes to master.

closes #145

### CODE REVIEW

File: .github/workflows/documentation.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+# Controls when the action will run. Triggers the workflow on push
+# events but only for the master branch
+on:
+  push:
+    branches:
+      - master
+      - feature/AutomaticDocDeployment # this will be removed once it's working
+jobs:
+  # This workflow contains a single job called "build"
```
Comment (iTitus) [inline .github/workflows/documentation.yml]: wrong comment

File: .github/workflows/documentation.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  gen_doc:
+    strategy:
+      matrix:
+        os: [ubuntu-18.04]
```
Comment (iTitus) [inline .github/workflows/documentation.yml]: why not `ubuntu-latest`?

File: .github/workflows/documentation.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        java: [11]
+    runs-on: ${{ matrix.os }}
+    name: asciidoctor to html
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+      - name: Load Cache
```
Comment (iTitus) [inline .github/workflows/documentation.yml]: please add a new line so one can see where one steps ends and another one begins.

File: .github/workflows/documentation.yml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+          restore-keys: |
+            ${{ runner.os }}-gradle-
+
+      - name: Set up JDK
+        uses: actions/setup-java@v1.3.0
+        with:
+          java-version: ${{ matrix.java }}
+
+      - name: Run Gradle test
+        run: ./gradlew test
```
Comment (iTitus) [inline .github/workflows/documentation.yml]: Not really necessary, but I know why you did it.
Let my objections be noted, but don't let them hinder the merge.

File: .github/workflows/documentation.yml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+      - name: Run Gradle Java Doc
+        run: ./gradlew javadoc
+
+      - name: Get build container
+        id: adocbuild
+        uses: avattathil/asciidoctor-action@master
+        with:
+          program: "asciidoctor -D doc -a data-uri --backend=html5 -o index.html doc/documentation.adoc"
+      - name: Print execution time
```
Comment (iTitus) [inline .github/workflows/documentation.yml]: new line, same reasoning as above

File: .github/workflows/documentation.yml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        run: echo "Time ${{ steps.adocbuild.outputs.time }}"
+      - name: Deploy docs to ghpages
```
Comment (iTitus) [inline .github/workflows/documentation.yml]: new line, same reasoning as above

File: .github/workflows/documentation.yml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        uses: peaceiris/actions-gh-pages@v3
+        with:
+          deploy_key: ${{ secrets.ACTIONS_DEPLOY_KEY }}
+          publish_dir: ./doc
```
Comment (iTitus) [inline .github/workflows/documentation.yml]: new line at end of file

File: .gitignore [hunk 8]
```diff
@@ -34,3 +34,5 @@ out/
 
 prod.env
 minio_data/
+/doc/api/generated-snippets/
+/doc/javadoc/
```
Comment (iTitus) [inline .gitignore]: not really necessary, as one should not run the javadoc task manually.
but not really important

File: README.md [hunk 9]
```diff
@@ -1,4 +1,4 @@
-![Gradle Checks](https://github.com/hhu-propra2/abschlussprojekt-mopse/workflows/Gradle%20Checks/badge.svg)
+![Gradle Checks](https://github.com/hhu-propra2/abschlussprojekt-mopse/workflows/Gradle%20Checks/badge.svg) ![Generate Documentation](https://github.com/hhu-propra2/abschlussprojekt-mopse/workflows/Generate%20Documentation/badge.svg)
```
Comment (iTitus) [inline README.md]: does the badge link work as you changed the procedure multiple times

File: .github/workflows/documentation.yml (hunk 1 above)
Comment (Segelzwerg) [reply .github/workflows/documentation.yml]: indeed

File: .github/workflows/documentation.yml (hunk 2 above)
Comment (Segelzwerg) [reply .github/workflows/documentation.yml]: Actually I don't know. It works with latest, but the [doc](https://github.com/peaceiris/actions-gh-pages#%EF%B8%8F-keeping-existing-files) of the tool says 18.04

File: .github/workflows/documentation.yml (hunk 3 above)
Comment (Segelzwerg) [reply .github/workflows/documentation.yml]: copy

File: .github/workflows/documentation.yml (hunk 4 above)
Comment (Segelzwerg) [reply .github/workflows/documentation.yml]: Actually we need it to generate the snippets from the Controller Tests.

File: .gitignore (hunk 8 above)
Comment (Segelzwerg) [reply .gitignore]: I just want to prevent accidental commits.

File: .github/workflows/documentation.yml (hunk 4 above)
Comment (iTitus) [reply .github/workflows/documentation.yml]: ah fair

File: README.md (hunk 9 above)
Comment (Segelzwerg) [reply README.md]: Yes I think so, but we could use as link to the documentation?

Comment (iTitus) [reply README.md]: the badge is fine, i just want to know if the link is still valid?
can you regenerate the badge link to be sure?

Comment (Segelzwerg) [reply README.md]: sure thing

File: .github/workflows/documentation.yml [hunk 10]
```diff
@@ -0,0 +1,53 @@
+# This is a workflow to generate the documentation automatically.
+
+name: Generate Documentation
+
+on:
+  push:
+    branches:
+      - master
+jobs:
```
Comment (iTitus) [inline .github/workflows/documentation.yml]:
```suggestion

jobs:
```

File: README.md (hunk 9 above)
Comment (Segelzwerg) [reply README.md]: regenerated it, same link as befor

File: .github/workflows/documentation.yml [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+      - name: Set up JDK
+        uses: actions/setup-java@v1.3.0
+        with:
+          java-version: ${{ matrix.java }}
+
+      - name: Run Gradle test
+        run: ./gradlew test
+
+      - name: Run Gradle Java Doc
```
Comment (iTitus) [inline .github/workflows/documentation.yml]:
```suggestion
      - name: Run Gradle Javadoc
```

File: README.md [hunk 12]
```diff
@@ -22,8 +22,8 @@ Es ist aufwendig für Lerngruppen oder Kurse Dateien miteinander zu teilen, wenn
 
 ## Dokumentation
 
-Hier kann die Dokumentation gefunden werden: [Dokumentation](https://hhu-propra2.github.io/abschlussprojekt-mopse/doc/documentation.html)
+Hier kann die Dokumentation gefunden werden: [Dokumentation](https://hhu-propra2.github.io/abschlussprojekt-mopse/)
 
 Javadoc kann hier aufgerufen werden: [Javadoc](https://hhu-propra2.github.io/abschlussprojekt-mopse/javadoc/)
 
-Die REST API Documentation befindet sich hier: [REST API](https://hhu-propra2.github.io/abschlussprojekt-mopse/api/external-api.html)
+Die REST API Documentation befindet sich hier: [REST API](https://hhu-propra2.github.io/abschlussprojekt-mopse/#section-system-scope-and-context)
```
Comment (iTitus) [inline README.md]: why the anchored link?

Comment (Segelzwerg) [reply README.md]: why not? So it is easier to find.

Comment (iTitus) [reply README.md]: the link should lead to the top

Comment (Segelzwerg) [reply README.md]: The documentation link leads to top.

Comment (iTitus) [reply README.md]: no it leads to the system-scope-and-context section, which is below the table of contents and maybe a preamble

**Pull request merged**